enterprise/backend/src/metabase_enterprise/metabot_v3/tools/api.clj

```diff
@@ -824,6 +958,64 @@
+  [_route-params
+   _query-params
+   {:keys [arguments conversation_id] :as body} :- [:merge
+                                                    [:map [:arguments {:optional true} ::search-data-sources-arguments]]
+                                                    ::tool-request]]
+  (metabot-v3.context/log (assoc body :api :search-data-sources) :llm.log/llm->be)
+  (try
+    (let [options (mc/encode ::search-data-sources-arguments
+                             arguments (mtx/transformer {:name :tool-api-request}))
+          {:keys [keywords description database-id entity_types limit]} options
+          ;; Default entity_types to all allowed values if not provided or empty
+          entity_types (if (seq entity_types)
+                         entity_types
+                         ["table" "model" "question" "dashboard" "metric"])
+          ;; Treat empty string description as nil
+          normalized-description (if (and (string? description) (str/blank? description)) nil description)
+          search-terms (concat (or keywords []) (when normalized-description [normalized-description]))
+          search-models (set (keep entity-type->search-model entity_types))
+          ;; Run a search for each keyword/description separately
+          all-results (->> search-terms
+                           (map (fn [term]
+                                  (let [search-context (search/search-context
+                                                        {:search-string term
+                                                         :models search-models
+                                                         :table-db-id database-id
+                                                         :current-user-id api/*current-user-id*
+                                                         ;;TODO: Do we need those impersonated, sandboxed, superuser checks?
+                                                         :is-impersonated-user? (perms/impersonated-user?)
+                                                         :is-sandboxed-user? (perms/sandboxed-user?)
+                                                         :is-superuser? api/*is-superuser?*
+                                                         :current-user-perms @api/*current-user-permissions-set*
+                                                         :context :default ;; TODO: Does this need to be a specific Metabot context maybe?
+                                                         :archived false
+                                                         :limit (or limit 50)
+                                                         :offset 0})
+                                        search-results (search/search search-context)]
+                                    (:data search-results))))
+                           (apply concat))
+          ;; Deduplicate by id/type
+          unique-results (vals (into {} (map (fn [r]
```

Comment:
`(vals (m/index-by (juxt :id :model) all-results))`

https://weavejester.github.io/medley/medley.core.html#var-index-by

Reply:
thank you